[asyncmix-master] Asynchronous mixing implementation

smkuls opened the pull request.

smkuls committed:
Implement asynchronous mixing

* Implement powermixing
* Implement butterfly network

amiller commented on tests/test_butterfly_network.py:
```diff
@@ -0,0 +1,26 @@
+    import honeybadgermpc.apps.shuffle.butterfly_network as butterfly
+    from honeybadgermpc.passive import generate_test_triples, Field, runProgramAsTasks
+
+    async def verify_output(ctx, **kwargs):
```
This test scenario looks great. It uses some preshared inputs, preshared -1/1 shares, run the butterfly network, and check that the output is indeed a permutation (the sorted arrays match).

sbellem committed:
Use pytest asyncio marker for all tests

Otherwise we somehow end up with two loops such that some tests run
twice, one run being successful meanwhile the other fails.

amiller commented on Dockerfile:
```diff
@@ -19,5 +18,14 @@ COPY . /usr/src/HoneyBadgerMPC
+# https://github.com/wbhart/flint2/issues/217
+# This has been fixed if we pull the latest code from the repo. However, we want
+# to avoid compiling the lib from the source since it adds 20 minutes to the build.
+RUN sed -i '30c #include "flint/flint.h"' /usr/include/flint/flintxx/flint_classes.h
```
ok good pragmatic save...
is there a way to make this assert that we're on the expected version of flint, would we notice if it that changed

smkuls replied:
I have actually opened an issue on the flint repo - https://github.com/wbhart/flint2/issues/512

I will follow up with them regarding this.

amiller reviewed (approved):
ok all this looks fine
